core/runner/src/main/java/com/cognifide/aet/runner/processing/data/SuiteDataService.java

```diff
@@ -15,38 +15,91 @@
  */
 package com.cognifide.aet.runner.processing.data;
 
+import static com.google.common.base.Strings.isNullOrEmpty;
+
 import com.cognifide.aet.communication.api.metadata.Suite;
 import com.cognifide.aet.communication.api.metadata.ValidatorException;
 import com.cognifide.aet.vs.MetadataDAO;
 import com.cognifide.aet.vs.SimpleDBKey;
 import com.cognifide.aet.vs.StorageException;
+import java.util.Optional;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 @Component(service = SuiteDataService.class)
 public class SuiteDataService {
 
+  private static final Logger LOGGER = LoggerFactory.getLogger(SuiteDataService.class);
+
   @Reference
   private MetadataDAO metadataDAO;
 
+  public SuiteDataService() {
+  }
+
+  public SuiteDataService(MetadataDAO metadataDAO) {
+    this.metadataDAO = metadataDAO;
+  }
+
   /**
    * @param currentRun - current suite run
-   * @return suite wrapper with all patterns from the last or specified (see
-   * Suite.patternCorrelationId) run, if this is the first run of the suite, patterns will be
-   * empty.
+   * @return suite wrapper with all patterns from the last or specified (see Suite.patternCorrelationId) run, if this is the first run of the suite, patterns will be empty.
    */
   public Suite enrichWithPatterns(final Suite currentRun) throws StorageException {
     final SimpleDBKey dbKey = new SimpleDBKey(currentRun);
     Suite lastVersion = metadataDAO.getLatestRun(dbKey, currentRun.getName());
-    Suite pattern;
-    if (currentRun.getPatternCorrelationId() != null) {
-      pattern = metadataDAO.getSuite(dbKey, currentRun.getPatternCorrelationId());
+    String checkSumCurrentRunProject = currentRun.getCheckSumProject();
+    final Suite pattern;
+
+    if (isTestRunWithCheckSum(checkSumCurrentRunProject)) {
+      Optional<Suite> suiteByChecksum = Optional.ofNullable(metadataDAO.getSuiteByChecksum(dbKey, checkSumCurrentRunProject));
+      pattern = suiteByChecksum.orElseGet(() -> assignCheckSumToPattern(dbKey, currentRun, lastVersion, checkSumCurrentRunProject));
     } else {
-      pattern = lastVersion;
+      if (currentRun.getPatternCorrelationId() == null) {
+        pattern = lastVersion;
+      } else {
+        pattern = metadataDAO.getSuite(dbKey, currentRun.getPatternCorrelationId());
+      }
     }
     return SuiteMergeStrategy.merge(currentRun, lastVersion, pattern);
   }
 
+  private Suite assignCheckSumToPattern(SimpleDBKey dbKey, Suite currentRun, Suite lastVersion, String checkSumCurrentRunProject) {
+    Suite pattern = null;
+    try {
+      if (isSuitHasPattern(currentRun)) {
```

Comment:
As far as I understand the code, if a current suite is provided with a `patternCorrelationId`, and with a current run `checksum`, this `checksum` will be assigned to the suite the`patternCorrelationId` is pointing to. 

Is this an expected behaviour? Due to this fact, multiple different codebases that are running tests with both these parameters passed will keep changing the `checksum` of the same suite in db. 

A situation could occur, when if a test would be run with only checksum parameter provided, the patterns may not be found, because some other environments could have changed the checksum, as they may have passed a `patternCorrelationId` of these patterns.